File: src/spectralnorm/spectralnorm.rs
fn eval_a(i: usize, j: usize) -> f64 {
    1.0 / (((i + j) * (i + j + 1) / 2 + i + 1) as f64)
}

fn eval_a_times_u(u: &[f64], au: &mut [f64]) {
    let n = u.len();
    for i in 0..n {
        au[i] = 0.0;
        for j in 0..n {
            au[i] += eval_a(i, j) * u[j];
        }
    }
}

fn eval_at_times_u(u: &[f64], au: &mut [f64]) {
    let n = u.len();
    for i in 0..n {
        au[i] = 0.0;
        for j in 0..n {
            au[i] += eval_a(j, i) * u[j];
        }
    }
}

fn eval_ata_times_u(u: &[f64], v: &mut [f64], w: &mut [f64]) {
    eval_a_times_u(u, w);
    eval_at_times_u(w, v);
}

fn main() {
    let n = std::env::args().nth(1).and_then(|x| x.parse().ok()).unwrap_or(5500);
    let mut u = vec![1.0; n];
    let mut v = vec![0.0; n];
    let mut w = vec![0.0; n];

    for _ in 0..10 {
        eval_ata_times_u(&u, &mut v, &mut w);
        eval_ata_times_u(&v, &mut u, &mut w);
    }

    let mut v_bv = 0.0;
    let mut vv = 0.0;
    for i in 0..n {
        v_bv += u[i] * v[i];
        vv += v[i] * v[i];
    }
    println!("{:.9}", (v_bv / vv).sqrt());
}

File: src/spectralnorm/spectralnorm.ml
let eval_A i j = 1.0 /. float_of_int ((i + j) * (i + j + 1) / 2 + i + 1)

let eval_A_times_u u au =
  let n = Array.length u in
  for i = 0 to n - 1 do
    au.(i) <- 0.0;
    for j = 0 to n - 1 do
      au.(i) <- au.(i) +. eval_A i j *. u.(j)
    done
  done

let eval_At_times_u u au =
  let n = Array.length u in
  for i = 0 to n - 1 do
    au.(i) <- 0.0;
    for j = 0 to n - 1 do
      au.(i) <- au.(i) +. eval_A j i *. u.(j)
    done
  done

let eval_AtA_times_u u v w =
  eval_A_times_u u w;
  eval_At_times_u w v

let () =
  let n = if Array.length Sys.argv > 1 then int_of_string Sys.argv.(1) else 5500 in
  let u = Array.make n 1.0 in
  let v = Array.make n 0.0 in
  let w = Array.make n 0.0 in
  for _i = 1 to 10 do
    eval_AtA_times_u u v w;
    eval_AtA_times_u v u w
  done;
  let vBv = ref 0.0 in
  let vv = ref 0.0 in
  for i = 0 to n - 1 do
    vBv := !vBv +. u.(i) *. v.(i);
    vv := !vv +. v.(i) *. v.(i)
  done;
  Printf.printf "%.9f\n" (sqrt (!vBv /. !vv))